SI-8553 WrappedArray throws exception on lastIndexWhere when index out of range

ruippeixotog opened the pull request:
Adds a check in `IndexedSeqOptimized#lastIndexWhere(A => Boolean, Int)` to begin searching in the end of the collection if `end` is greater than the collection's length.

Discussed in https://groups.google.com/d/topic/scala-internals/-MacXivbY0Q/discussion.

kiritsuku commented on src/library/scala/collection/IndexedSeqOptimized.scala:
```diff
@@ -206,7 +206,7 @@ trait IndexedSeqOptimized[+A, +Repr] extends Any with IndexedSeqLike[A, Repr] {
   override /*SeqLike*/
   def lastIndexWhere(p: A => Boolean, end: Int): Int = {
-    var i = end
+    var i = end min self.length - 1
```
`math.min` is more efficient. Though, not sure if that is very important here.

ruippeixotog replied:
I followed the code style used in `indexWhere` just above this method, where `from max 0` was used. If you think it is important, I can change the both methods to use `math.min` and `math.max` respectively.

som-snytt replied:
Both styles are used in the file. (i min j is so easy to read, at a cost.)  But `self` alias is used only as necessary; compare `this(i) == that(j)` or other uses of `length`.  When I see the alias, my eye has to look for the enclosing object.

I don't have a deep, experience-based opinion about this, but it's too bad to calculate length if not necessary. (Maybe it's only theoretical: loading an array of records a chunk at a time, and length forces loading all.)  It's easy to say `try { ... } catch { case _: IndexOutOfBoundsException => -1 }`; that is not entirely free of cost, but it mostly just punishes the error case.

Wow, I just tried "edit in fullscreen Zen mode", that's just great.  I think I want my code editor to offer that.

ruippeixotog replied:
Ok, you're right about `min`, `max` and `self` (I actually prefer to use `math.min` in my own code, I only used `RichInt`'s `min` because I saw in `indexWhere` and wanted to follow the style). I'm going to change them.

I'm not so sure about changing `length` usage to a `try`, as I'm afraid it will bring some penalties to performance (and to readability, of course). Moreover, [IndexedSeqOptimized](http://www.scala-lang.org/api/2.11.0/index.html#scala.collection.IndexedSeqOptimized) scaladoc states that "Indexed sequences support constant-time or near constant-time element access and length computation. They are defined in terms of abstract methods apply for indexing and length". From the point of view of a guy who must implement a custom collection, I would assume from there that I must provide a efficient `length` if I want to get the benefits of `IndexSeqOptimized`. I see `length` being freely used throughout the class even when not explicitly needed, such as in `startsWith`, `segmentLength` or even in `isEmpty`.

This is the first pull request that I'm submitting to Scala, so I don't know how should I proceed. Can I add a new commit to this pull request with the fixes you mentioned? Or should I close this PR and open a new one containing a single commit already with the changes?

som-snytt replied:
OK, you're totally right that the doc promises length is fast.  Sorry I didn't read carefully.

You can `push -f` a squashed commit.  (Or a new branch and PR if you want to preserve the old history for some reason.)

som-snytt commented on test/junit/scala/collection/IndexedSeqOptimizedTest.scala:
```diff
@@ -0,0 +1,15 @@
+  @Test
+  def notThrowsAnExceptionInLastIndexOf() {
+    assert((Array(2): collection.mutable.WrappedArray[Int]).lastIndexWhere(_ => true, 1) == 0)
+    assert("abc123".lastIndexWhere(_.isLetter, 6) == 2)
```
I don't know if there's a policy, because usage is still mixed, but I'm trying to be good and

```
assertEquals(2, "abc123".lastIndexWhere(_.isLetter, 6))
```

though I'm unhappy with the lack of smart asserts.  It's like Scala was born ten years ago, so its new unit tests must dress in period costume.

ruippeixotog replied:
I fully agree with you on the usage of rich asserts. I used `assert` because almost all of the JUnit tests in the `scala.collection` package use `Predef#assert`. As I'm new to contributing to Scala and I'm not sure of what is a policy and what is not, I just blended the code in :)

ruippeixotog committed:
SI-8553 WrappedArray throws exception on lastIndexWhere when index out of range

Adds a check in `IndexedSeqOptimized#lastIndexWhere(A => Boolean, Int)`
to begin searching in the end of the collection if `end` is greater than
the collection's length.

Discussed in
https://groups.google.com/d/topic/scala-internals/-MacXivbY0Q/discussion.

ruippeixotog commented:
@som-snytt fixed it right now.

ruippeixotog commented:
Review by @Ichoran

lrytz commented:
ping @Ichoran for review

Ichoran commented:
LGTM.
